src/EditorFeatures/Core/Implementation/BlockCommentCompletion/AbstractBlockCommentCompletionCommandHandler.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Editor.Commands;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.BlockCommentCompletion
+{
+    internal abstract class AbstractBlockCommentCompletionCommandHandler : ICommandHandler<ReturnKeyCommandArgs>
+    {
+        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
+        private readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        protected AbstractBlockCommentCompletionCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            Contract.ThrowIfNull(undoHistoryRegistry);
+            Contract.ThrowIfNull(editorOperationsFactoryService);
+
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public CommandState GetCommandState(ReturnKeyCommandArgs args, Func<CommandState> nextHandler) => nextHandler();
+
+        public void ExecuteCommand(ReturnKeyCommandArgs args, Action nextHandler)
+        {
+            var subjectBuffer = args.SubjectBuffer;
+            var textView = args.TextView;
+
+            var caretPosition = textView.GetCaretPoint(subjectBuffer);
+            if (caretPosition == null)
+            {
+                nextHandler();
+                return;
```

Comment:
a *slightly* nicer pattern for me would be:

```c#
public void Execute(...)
{
    if (TryHandleReturnKey(...))
    {
        // Don't let any other command handler run, we completely handled the return key.
        return;
    }

    // Defer to the next command handler to process this.
    nextHandler();
}

private bool TryHandleReturnKey(...)
{
    .. your code ...
}
```

This now puts the job of command handler chain management in a single place.  It means also if someone adds any sort of control flow here that they have to worry about forgetting to do this.